14/add organization authentication

AdamVig opened the pull request:
Closes #14.

This PR will cause test coverage to decline. #54 addresses test coverage for this feature.

AdamVig committed:
Move test database instance into fixture module

This allows more than one test to use the database instance without
duplicating code between tests.

AdamVig committed:
Add environment variables to database service

This allows for importing the database service without first importing
`index.js`. `index.js` takes care of initializing the environment
variables when the app is run normally, but in a test environment,
individual controllers may use the service, and controllers do not
initialize the application themselves.

AdamVig committed:
Add methods to authentication controller

- Add environment variable for storing JWT secret
- Add authentication controller with functions to generate token,
  authenticate token, register and organization, and initialize Passport
- Rename `authentication.js` to `auth.js` for brevity
- Add tests for partial functionality of authentication controller

AdamVig committed:
Add authentication to all routes

Now, a request to any route except `/` must have the header
`Authorization` with the value `Bearer ` followed by a JWT. The JWT must
have a payload including `sub`, where `sub` (subject) is the ID of an
organization. This will be decoded in each request in order to help the
next handler in the endpoint handler chain which organization to look
for (this is important because data for all organizations is stored in
the same tables).

AdamVig committed:
Fix handlers not returning callback functions

It is a useful pattern to return callback functions such as `next()`,
because it ensures that nothing else will happen in the calling function
after that callback.

AdamVig committed:
Fix Travis config for bcrypt install

The Travis build was failing to compile bcrypt due to a C++
incompatibility issue. See [https://goo.gl/BS3stK](the Travis docs) for
the source of this change.

emroussel commented on controllers/auth.js:
```diff
@@ -0,0 +1,97 @@
+
+// Register an organization
+module.exports.register = (req, res, next) => {
+  if (req.body.name && req.body.email && req.body.password) {
```
Should there be a check after this to make sure the email is not already used and send an appropriate error message if it is the case?

AdamVig replied:
Good point. I think I might add a unique constraint to the email column in the database, ensuring that an error will be returned if an email is non-unique. What do you think?

emroussel replied:
Yeah good idea. Would that return the appropriate error code from the api though? (I think it should be a 409)

AdamVig replied:
Good point! I'll catch the error from the database and return the appropriate error code.

AdamVig replied:
Just want to point out that this functionality already exists, so I won't actually have to add a commit to this PR in order for it to work (I outsmarted myself): [see db.js](https://github.com/AdamVig/stockpile-api/blob/master/services/db.js#L26). 😃

emroussel reviewed (changes requested):
Can you add a section in the README describing how to add the authentication header to requests (mostly for the format)?

AdamVig committed:
Add token to auth response

The `/auth` endpoint was not returning a token, which is the entire
purpose of the endpoint. Its response is in the same format as
`/register`.

AdamVig committed:
Add API documentation on authentication

This will serve as a temporary solution until actual API documentation
is set up and hosted.
